pkg/export/setup/setup.go

```diff
@@ -0,0 +1,160 @@
+		Default(opts.ProjectID).StringVar(&opts.ProjectID)
+
+	// The location and cluster flag should probably not be used. On the other hand, they make it easy
+	// to populate these important values in the monitored resource without interfering with existing
+	// Prometheus configuration.
+	a.Flag("export.label.location", fmt.Sprintf("The default location set for all exported data. Prefer setting the external label %q in the Prometheus configuration if not using the auto-discovered default.", export.KeyLocation)).
+		Default(opts.Location).StringVar(&opts.Location)
+
+	a.Flag("export.label.cluster", fmt.Sprintf("The default cluster set for all scraped targets. Prefer setting the external label %q in the Prometheus configuration if not using the auto-discovered default.", export.KeyCluster)).
+		Default(opts.Cluster).StringVar(&opts.Cluster)
+
+	a.Flag("export.match", `A Prometheus time series matcher. Can be repeated. Every time series must match at least one of the matchers to be exported. This flag can be used equivalently to the match[] parameter of the Prometheus federation endpoint to selectively export data. (Example: --export.match='{job="prometheus"}' --export.match='{__name__=~"job:.*"})`).
+		Default("").SetValue(&opts.Matchers)
+
+	a.Flag("export.debug.metric-prefix", "Google Cloud Monitoring metric prefix to use.").
+		Default(export.MetricTypePrefix).StringVar(&opts.MetricTypePrefix)
+
+	a.Flag("export.debug.disable-auth", "Disable authentication (for debugging purposes).").
+		Default("false").BoolVar(&opts.DisableAuth)
+
+	a.Flag("export.debug.batch-size", "Maximum number of points to send in one batch to the GCM API.").
+		Default(strconv.Itoa(export.BatchSizeMax)).UintVar(&opts.BatchSize)
+
+	haBackend := a.Flag("export.ha.backend", fmt.Sprintf("Which backend to use to coordinate HA pairs that both send metric data to the GCM API. Valid values are %q or %q", HABackendNone, HABackendKubernetes)).
+		Default(HABackendNone).Enum(HABackendNone, HABackendKubernetes)
+
+	kubeConfigPath := a.Flag("export.ha.kube.config", "Path to kube config file.").
+		Default("").String()
+	kubeNamespace := a.Flag("export.ha.kube.namespace", "Namespace for the HA locking resource. Must be identical across replicas. May be set through the KUBE_NAMESPACE environment variable.").
+		Default("").OverrideDefaultFromEnvar("KUBE_NAMESPACE").String()
+	kubeName := a.Flag("export.ha.kube.name", "Name for the HA locking resource. Must be identical across replicas. May be set through the KUBE_NAME environment variable.").
+		Default("").OverrideDefaultFromEnvar("KUBE_NAME").String()
+
+	return func(logger log.Logger, metrics prometheus.Registerer) (*export.Exporter, error) {
+		switch *haBackend {
+		case HABackendNone:
+		case HABackendKubernetes:
+			kubecfg, err := loadKubeConfig(*kubeConfigPath)
+			if err != nil {
+				return nil, errors.Wrap(err, "loading kube config failed")
+			}
+			opts.HighAvailabilityRange, err = lease.NewKubernetes(
```

Comment:
I agree it makes the whole thing a good bit more awkward.
But instantiating it directly means `export` will depend on Kubernetes (and later GCS) client libraries, which means insane amounts of code. That's what I mainly wanted to avoid.

I'd not focus too much on the flag<>opts coupling, which is why I factored it into a setup package to begin with.

Reply:
Makes sense. However, if `lease` is moved into `export`, the dependency is unavoidable, right?

Reply:
Directory structure does not imply build-time dependency.